stringify.c++ compile error with vs2015 update3

line 186 "return print(element, which, indent.next(), BARE);"
1>..\..\..\src\capnp\stringify.cpp(186): error C2664: 'kj::StringTree capnp::`anonymous-namespace'::print(const capnp::DynamicValue::Reader &,capnp::schema::Type::Which,capnp::`anonymous-namespace'::Indent,capnp::`anonymous-namespace'::PrintMode)': cannot convert argument 1 from 'kj::_::Mapper<T>::Element' to 'const capnp::DynamicValue::Reader &'
1>          with
1>          [
1>              T=capnp::DynamicList::Reader &
1>          ]
1>  ..\..\..\src\capnp\stringify.cpp(186): note: Reason: cannot convert from 'kj::_::Mapper<T>::Element' to 'const capnp::DynamicValue::Reader'
1>          with
1>          [
1>              T=capnp::DynamicList::Reader &
1>          ]
1>  ..\..\..\src\capnp\stringify.cpp(186): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
1>..\..\..\src\capnp\stringify.cpp(187): error C2678: binary '*': no operator found which takes a left-hand operand of type 'kj::_::Mapper<T>' (or there is no acceptable conversion)
1>          with
1>          [
1>              T=capnp::DynamicList::Reader &
1>          ]
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(800): note: could be 'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicList::Builder &)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(799): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicList::Reader &)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(798): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicStruct::Builder &)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(797): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicStruct::Reader &)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(796): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,capnp::DynamicEnum)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(795): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicValue::Builder &)'
1>  c:\work\capnproto-master\cpp\src\capnp\dynamic.h(794): note: or       'kj::StringTree capnp::operator *(kj::_::Stringifier,const capnp::DynamicValue::Reader &)'
1>  c:\work\capnproto-master\cpp\src\capnp\schema.h(317): note: or       'kj::StringPtr capnp::operator *(kj::_::Stringifier,const capnp::StructSchema::Field &)'
1>  c:\work\capnproto-master\cpp\src\capnp\schema.h(185): note: or       'kj::StringPtr capnp::operator *(kj::_::Stringifier,const capnp::Schema &)'
1>  c:\work\capnproto-master\cpp\src\capnp\any.h(83): note: or       'kj::StringPtr capnp::operator *(kj::_::Stringifier,capnp::Equality)'
1>  c:\work\capnproto-master\cpp\src\capnp\blob.h(173): note: or       'kj::StringPtr capnp::operator *(kj::_::Stringifier,capnp::Text::Builder)'
1>  c:\work\capnproto-master\cpp\src\capnp\common.h(64): note: or       'kj::StringPtr capnp::operator *(kj::_::Stringifier,capnp::Void)'
1>  ..\..\..\src\capnp\stringify.cpp(187): note: while trying to match the argument list '(kj::_::Mapper<T>, capnp::`anonymous-namespace'::print::<lambda_6ec554b1b68b306ff2fb79943320ece0>)'
1>          with
1>          [
1>              T=capnp::DynamicList::Reader &
1>          ]

sorry, I made a wrong description about the error.
At first, there were two errors:
1. c:\work\capnproto-master\c++\src\kj\array.h(527): error C2228: left of '.begin' must have class/struct/union

2. C:\work\capnproto-master\c++\src\capnp\stringify.c++(187): error C2678: binary '*': no operator found which takes a left-hand operand of type 'kj::_::Mapper<T>' (or there is no acceptable conversion)
1>          with
1>          [
1>              T=capnp::DynamicList::Reader &
1>          ]
...

For the first error, I change the source code from "typedef decltype(*(array).begin()) Element;" to "typedef decltype((*(array)).begin()) Element;". Now, the first error is disappear, but the second error still exists and contents changed -- they are the errors I describe in my source post.

--- a/c++/src/kj/array.h
+++ b/c++/src/kj/array.h
@@ -524,7 +524,7 @@ struct Mapper {
     }
     return builder.finish();
   }
-  typedef decltype(*(array).begin()) Element;
+  typedef decltype(*kj::instance<T>().begin()) Element;
 };
 
 template <typename T, size_t s>

--- a/c++/src/kj/async-io.c++
+++ b/c++/src/kj/async-io.c++
@@ -122,7 +122,7 @@ private:
     auto partPtr = part.asPtr();
     parts.add(kj::mv(part));
     return input.tryRead(partPtr.begin(), partPtr.size(), partPtr.size())
-        .then([this,partPtr,total](size_t amount) -> Promise<uint64_t> {
+        .then([this,KJ_CPCAP(partPtr),total](size_t amount) -> Promise<uint64_t> {
       uint64_t newTotal = total + amount;
       if (amount < partPtr.size()) {
         return newTotal;

